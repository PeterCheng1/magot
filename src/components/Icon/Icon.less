@import '../../variable.less';
@import '../../iconfont/iconfont.less';

.@{prefix}-icon {
  .box-sizing();
  .box();

  display: inline-block;
  font-style: normal;
  text-align: center;
  text-rendering: optimizeLegibility;
  text-indent: 0;

  &.spin {
    animation: mgt-icon-spin @icon-spin-duration infinite linear;
  }
}

@keyframes mgt-icon-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
