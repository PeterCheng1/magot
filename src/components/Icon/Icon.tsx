import React from 'react';

import './Icon.less';

import * as component from '../component';

export interface IconProps extends component.ComponentBase {
  /**
   * 图标名称，即iconfont中的图标名
   */
  name: string;

  /**
   * 是否旋转图标
   * @default false
   */
  spin?: boolean;

  /**
   * 图标大小，即iconfont的font-size值
   */
  size?: number;
}

const defaultProps: Partial<IconProps> = {
  spin: false,
};

function Icon(props: IconProps) {
  const type = 'icon';
  const prefix = component.getComponentPrefix(type);
  const cls = component.getComponentClasses(
    type,
    { ...props, size: 'normal' },
    `${prefix}-${props.name}`,
    {
      spin: !!props.spin,
    }
  );
  return <i className={cls} style={{ ...props.style, fontSize: props.size }} />;
}

Icon.defaultProps = defaultProps;

export default Icon;
