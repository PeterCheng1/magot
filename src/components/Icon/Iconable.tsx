import React from 'react';
import Icon from './Icon';
import * as component from '../component';

export type IconPosition = 'left' | 'right';

export interface IconableProps extends component.NestedComponent {
  /**
   * 图标
   */
  name?: string;

  /**
   * 是否旋转图标
   */
  spin?: boolean;

  /**
   * 图标位置，可选值有：`left`, `right`，默认`left`。
   * @default left
   */
  position?: IconPosition;

  /**
   * 图标大小，即`Icon`组件的`fontSize`属性
   */
  size?: number;
}

const defaultProps: Partial<IconableProps> = {
  position: 'left',
};

function Iconable(props: IconableProps) {
  const children = [props.children];
  let icon = props.name ? <Icon key="1" {...props} /> : null;
  if (props.position === 'right') {
    children.push(icon);
  } else {
    children.unshift(icon);
  }
  return <>{children}</>;
}

Iconable.defaultProps = defaultProps;

export default Iconable;
