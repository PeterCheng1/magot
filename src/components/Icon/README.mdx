---
name: Icon
route: /components/Icon
menu: Components
---

import { Playground, PropsTable } from 'docz';
import Icon from './Icon';

# Icon

矢量图标组件，用于标记带有特定含义的图标描述场景。

## Usage

<Playground>
  <Icon name="scan" />
  <Icon name="search" />
  <Icon name="download" />
  <Icon name="loading" spin={true} />
  <Icon name="chat" style={{ color: '#1890ff', fontSize: 32 }} />
</Playground>

## Props

<PropsTable of={Icon} />
