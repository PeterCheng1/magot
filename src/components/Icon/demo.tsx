import React from 'react';
import Icon from './Icon';
import { Link } from 'react-router-dom';

const icons = __ICONS__ || [];

function IconDemo() {
  return (
    <>
      <Link to="/" className="demo-goback">
        返回
      </Link>
      <div className="demo-box">
        {icons.map(icon => {
          return (
            <Icon
              key={icon}
              name={icon}
              size={32}
              style={{
                color: '#bfbfbf',
                marginRight: 30,
                marginBottom: 30,
              }}
            />
          );
        })}
      </div>
    </>
  );
}

export default IconDemo;
