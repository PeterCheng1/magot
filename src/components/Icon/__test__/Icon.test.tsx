import React from 'react';
import { render } from 'enzyme';
import Icon from '../Icon';
import {
  injectStyleTestSuites,
  injectClickEventTestSuite,
} from '../../../../test/shared.test';

describe('Icon', () => {
  it('should renders dom correctly', () => {
    const wrapper = render(<Icon />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should renders `.mgt-icon` className', () => {
    const wrapper = render(<Icon />);
    expect(wrapper.hasClass('mgt-icon')).toEqual(true);
  });

  it('should renders `name` prop when passed in', () => {
    const wrapper = render(<Icon name="home" />);
    expect(wrapper.hasClass('mgt-icon-home')).toEqual(true);
  });

  it('should renders `spin` prop when passed in', () => {
    const wrapper = render(<Icon spin={true} />);
    expect(wrapper.hasClass('spin')).toEqual(true);
  });

  it('should renders `size` prop when passed in', () => {
    const wrapper = render(<Icon size={28} />);
    expect(wrapper).toMatchSnapshot();
  });

  injectStyleTestSuites(Icon);
  injectClickEventTestSuite(Icon);
});
