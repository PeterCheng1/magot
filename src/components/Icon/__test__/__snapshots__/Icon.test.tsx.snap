// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon should renders \`size\` prop when passed in 1`] = `
<i
  class="mgt-icon"
  style="font-size:28px"
/>
`;

exports[`Icon should renders custom className 1`] = `
<i
  class="mgt-icon jest-test-class"
/>
`;

exports[`Icon should renders custom styles 1`] = `
<i
  class="mgt-icon"
  style="color:#333;background-color:#f0f0f0"
/>
`;

exports[`Icon should renders dom correctly 1`] = `
<i
  class="mgt-icon"
/>
`;
